homeassistant/components/lastfm/config_flow.py

```diff
@@ -0,0 +1,227 @@
+    """Get lastFM User."""
+    return LastFMNetwork(api_key=api_key).get_user(username)
+
+
+def validate_lastfm_user(user: User) -> dict[str, str] | None:
+    """Return error if the user is not correct. None if it is correct."""
+    errors = {}
+    try:
+        user.get_playcount()
+    except WSError as error:
+        LOGGER.error(error)
+        if error.details == "User not found":
+            errors["base"] = "invalid_account"
+        elif (
+            error.details
+            == "Invalid API key - You must be granted a valid key by last.fm"
+        ):
+            errors["base"] = "invalid_auth"
+        else:
+            errors["base"] = "unknown"
+    except Exception:  # pylint:disable=broad-except
+        errors["base"] = "unknown"
+    if not errors:
+        return None
+    return errors
+
+
+class LastFmConfigFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Config flow handler for LastFm."""
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        self.data: dict[str, Any] = {}
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> LastFmOptionsFlowHandler:
+        """Get the options flow for this handler."""
+        return LastFmOptionsFlowHandler(config_entry)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Initialize user input."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.data[CONF_API_KEY] = user_input[CONF_API_KEY]
+            self.data[CONF_MAIN_USER] = user_input[CONF_MAIN_USER]
+            main_user = get_lastfm_user(
+                self.data[CONF_API_KEY], self.data[CONF_MAIN_USER]
+            )
+            lastfm_errors = validate_lastfm_user(main_user)
+            if not lastfm_errors:
+                return await self.async_step_friends()
+            errors = lastfm_errors
+        return self.async_show_form(
+            step_id="user",
+            errors=errors,
+            description_placeholders=PLACEHOLDERS,
+            data_schema=self.add_suggested_values_to_schema(CONFIG_SCHEMA, user_input),
+        )
+
+    async def async_step_friends(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Form to select other users and friends."""
+        errors = {}
+        valid_users = []
+        if user_input is not None:
+            for username in user_input[CONF_USERS]:
+                lastfm_user = get_lastfm_user(self.data[CONF_API_KEY], username)
+                lastfm_errors = validate_lastfm_user(lastfm_user)
+                if lastfm_errors:
+                    errors = lastfm_errors
+                else:
+                    valid_users.append(username)
+            if not errors:
+                return self.async_create_entry(
+                    title="LastFM",
+                    data={
```

Comment:
```suggestion
                    data={},
                    options={
```

Reply:
What is the difference between data and options? I saw both used but I didn;t get when to use what.

Reply:
The general idea is that `data` should be persistent (for example) your `api_key` and `main_user` while the `users_list` that can later be changed in options flow should go in options.

Reply:
Does this also go for setting up new config entries via UI? I haven't really looked into it, but I can't remember integrations actively splitting their data into data and options

Reply:
For `async_create_entry` in `ConfigFlow` you need to separate `data` and `options`.
For `async_create_entry` in `OptionsFlow` what you put in `data` is saved in `options`

Some helpers store into options regardless as it's easier to handle from `OptionsFlow` but in minimum what can be changed in `OptionsFlow` should go in `options`.

Reply:
Thanks for the explanation!

Reply:
But that does mean that we should split API_KEY and MAIN_USER from the USERS and passthem both in their own fields right?